craft_parts/sources/sources.py

```diff
@@ -71,16 +71,113 @@
 cases you want to refer to the documentation for the specific plugin.
 """
 
-from typing import Dict, Type
+import os
+import re
+from typing import TYPE_CHECKING, Dict, Optional, Type
 
+from craft_parts.dirs import ProjectDirs
+
+from . import errors
 from .base import SourceHandler
 from .local_source import LocalSource
 from .tar_source import TarSource
 
+if TYPE_CHECKING:
+    from craft_parts.parts import Part
+
 SourceHandlerType = Type[SourceHandler]
 
 
 _source_handler: Dict[str, SourceHandlerType] = {
     "local": LocalSource,
     "tar": TarSource,
 }
+
+
+def get_source_handler(
+    application_name: str,
+    part: "Part",
+    dirs: ProjectDirs,
+) -> Optional[SourceHandler]:
+    """Return the appropriate handler for the given source.
+
+    :param application_name: The name of the application using Craft Parts.
+    :param part: The part to get a source handler for.
+    :param dirs: The project's work directories.
+    """
+    if not dirs:
+        dirs = ProjectDirs()
+
+    source_handler = None
+    if part.spec.source:
+        handler_class = _get_source_handler_class(
+            part.spec.source,
+            source_type=part.spec.source_type,
+        )
+        source_handler = handler_class(
+            application_name=application_name,
+            source=part.spec.source,
+            part_src_dir=part.part_src_dir,
+            source_checksum=part.spec.source_checksum,
+            source_branch=part.spec.source_branch,
+            source_tag=part.spec.source_tag,
+            source_depth=part.spec.source_depth,
+            source_commit=part.spec.source_commit,
+            dirs=dirs,
+        )
+
+    return source_handler
+
+
+def _get_source_handler_class(source, *, source_type: str = "") -> SourceHandlerType:
```

Comment:
on load is source_type None or ""?

Reply:
It's an empty string in the unmarshaled part if not specified.